src/test/scala/io/iohk/ethereum/network/discovery/PeerDiscoveryManagerSpec.scala

```diff
@@ -0,0 +1,223 @@
+    new Fixture {
+      override lazy val discoveryConfig =
+        defaultConfig.copy(discoveryEnabled = false, reuseKnownNodes = false, bootstrapNodes = Set.empty)
+
+      override def test(): Unit = {
+        getPeers.futureValue.nodes shouldBe empty
+      }
+    }
+  }
+
+  it should "serve the bootstrap nodes if known peers are reused even discovery isn't enabled and the manager isn't started" in test {
+    new Fixture {
+      override lazy val discoveryConfig =
+        defaultConfig.copy(discoveryEnabled = false, reuseKnownNodes = true, bootstrapNodes = sampleBootstrapNodes)
+
+      override def test(): Unit = {
+        getPeers.futureValue.nodes should contain theSameElementsAs (sampleBootstrapNodes)
+      }
+    }
+  }
+
+  it should "serve the known peers if discovery is enabled and the manager isn't started" in test {
+    new Fixture {
+      override lazy val discoveryConfig =
+        defaultConfig.copy(discoveryEnabled = true, reuseKnownNodes = true, bootstrapNodes = Set.empty)
+
+      (knownNodesStorage.getKnownNodes _)
+        .expects()
+        .returning(sampleKnownUris)
+        .once()
+
+      override def test(): Unit = {
+        getPeers.futureValue.nodes.map(_.toUri) should contain theSameElementsAs (sampleKnownUris)
+      }
+    }
+  }
+
+  it should "merge the known peers with the service if it's started" in test {
+    new Fixture {
+      override lazy val discoveryConfig =
+        defaultConfig.copy(discoveryEnabled = true, reuseKnownNodes = true, bootstrapNodes = Set.empty)
+
+      val sampleNodes = sampleBootstrapNodes
+
+      (knownNodesStorage.getKnownNodes _)
+        .expects()
+        .returning(sampleKnownUris)
+        .once()
+
+      (discoveryService.getNodes _)
+        .expects()
+        .returning(Task(sampleNodes.map(toENode)))
+        .once()
+
+      val expected = sampleKnownUris ++ sampleNodes.map(_.toUri)
+
+      override def test(): Unit = {
+        peerDiscoveryManager ! PeerDiscoveryManager.Start
+        eventually {
+          getPeers.futureValue.nodes.map(_.toUri) should contain theSameElementsAs (expected)
+        }
+      }
+    }
+  }
+
+  it should "keep serving the known peers if the service fails to start" in test {
+    new Fixture {
+      override lazy val discoveryConfig =
+        defaultConfig.copy(discoveryEnabled = true, reuseKnownNodes = true, bootstrapNodes = Set.empty)
+
+      override lazy val discoveryServiceResource: Resource[Task, DiscoveryService] =
+        Resource.liftF(Task.raiseError[DiscoveryService](new RuntimeException("Oh no!") with NoStackTrace))
+
+      (knownNodesStorage.getKnownNodes _)
+        .expects()
+        .returning(sampleKnownUris)
+        .once()
+
+      override def test(): Unit = {
+        peerDiscoveryManager ! PeerDiscoveryManager.Start
+        Thread.sleep(100)
+        getPeers.futureValue.nodes should have size (sampleKnownUris.size)
```

Comment:
Couldn't we remove the sleep by having an eventually here?

Reply:
I'll have a think.. the problem is that the starting is happening in the background and I'm testing that a failed start doesnt' cause any problems, when looking from the outside.

Reply:
I'll set some variable that we can wait on.